src/renderers/shared/stack/reconciler/ReactCompositeComponent.js

```diff
@@ -1107,7 +1107,9 @@ var ReactCompositeComponent = {
     if (__DEV__) {
       var componentName = component && component.getName ?
         component.getName() : 'a component';
-      warning(publicComponentInstance != null,
+      warning(
+        publicComponentInstance != null ||
+        component._compositeType !== CompositeTypes.StatelessFunctional,
```

Comment:
`component._currentElement` is still polymorphic. We'd probably need to pass the component type along with the component instance if we wanted to avoid this.